ocaml/doc/pdf/api_ref_intro.md

```diff
@@ -0,0 +1,485 @@
+
+The API is presented here as a set of Remote Procedure Calls (RPCs), with a wire
+format based upon XML-RPC. No specific language bindings are prescribed,
+although examples will be given in the python programming language.
+
+Although we adopt some terminology from object-oriented programming,
+future client language bindings may or may not be object oriented.
+The API reference uses the terminology _classes_ and _objects_.
+For our purposes a _class_ is simply a hierarchical namespace;
+an _object_ is an instance of a class with its fields set to
+specific values. Objects are persistent and exist on the server-side.
+Clients may obtain opaque references to these server-side objects and then
+access their fields via get/set RPCs.
+
+For each class we specify a list of
+fields along with their _types_ and _qualifiers_. A qualifier is one of:
+
+* _RO~run~_: the field is Read Only. Furthermore, its value is
+  automatically computed at runtime.
+  For example, current CPU load and disk IO throughput.
+
+* _RO~ins~_: the field must be manually set when a new object is
+  created, but is then Read Only for the duration of the object's life.
+  For example, the maximum memory addressable by a guest is set
+  before the guest boots.
+
+* _RW_: the field is Read/Write. For example, the name of a VM.
+
+A full list of types is given in chapter [API Reference](#api-reference).
+However, there are three types that require explicit mention:
+
+* _t Ref_: signifies a reference to an object of type _t_.
+* _t Set_: signifies a set containing values of type _t_.
+* _(t~1~, t~2~) Map_: signifies a mapping from values of
+  type _t~1~_ to values of type _t~2~_.
+
+Note that there are a number of cases where _Refs_ are _doubly linked_.
+For example, a `VM` has a field called `VIFs` of type _(VIF Ref) Set_;
+this field lists the network interfaces attached to a particular VM.
+Similarly, the VIF class has a field called `VM` of type _VM Ref_
+which references the VM to which the interface is connected.
+These two fields are _bound together_, in the sense that
+creating a new VIF causes the `VIFs` field of the corresponding
+VM object to be updated automatically.
+
+The chapter [API reference](#api-reference) explicitly lists the fields that are
+bound together in this way. It also contains a diagram that shows
+relationships between classes. In this diagram an edge signifies the
+existence of a pair of fields that are bound together, using standard
+crows-foot notation to signify the type of relationship (e.g.
+one-many, many-many).
+
+## RPCs associated with fields
+
+Each field, _f_, has an RPC accessor associated with it
+that returns _f_'s value:
+
+* `"get_f(Ref x)"`: takes a _Ref_ that refers to an object and returns the value
+  of _f_.
+
+Each field, _f_, with attribute _RW_ and whose outermost type is _Set_ has the
+following additional RPCs associated with it:
+
+* `"add_to_f(Ref x, v)"`: adds a new element _v_ to the set referred to by
```

Comment:
I retained this from the original pdf; I think that was originally added in order to distinguish the real call names (e.g. add_to_guest_agent_config) from pattern call names (add_to_f). I agree it's a bit ugly, so I can remove the quotes if it's obvious what it is.